pyomo/core/plugins/transform/scaling.py

```diff
@@ -0,0 +1,269 @@
+            for k in component:
+                c = component[k]
+                # perform the constraint/objective scaling and variable sub
+                scaling_factor = component_scaling_factor_map[c]
+                if isinstance(c, _ConstraintData):
+                    _substitute_var_in_constraint(c, variable_substitution_map)
+
+                    if c._lower is not None:
+                        c._lower = c._lower * scaling_factor
+                    if c._upper is not None:
+                        c._upper = c._upper * scaling_factor
+
+                    if scaling_factor < 0:
+                        # swap inequalities
+                        temp = c._lower
+                        c._lower = c._upper
+                        c._upper = temp
+
+                    c._body = c._body * scaling_factor
+                    
+                    if scale_constraint_dual and c in model.dual:
+                        dual_value = model.dual[c]
+                        if dual_value is not None:
+                            model.dual[c] = dual_value/scaling_factor
+                    
+                elif isinstance(c, _ObjectiveData):
+                    _substitute_var(c.expr, variable_substitution_map)
+                    c.expr = c.expr * scaling_factor
+                else:
+                    raise NotImplementedError('Unknown object type found when applying scaling factors in ScaleModel transformation - Internal Error')
+                
+        model.component_scaling_factor_map = component_scaling_factor_map
+
+        return model
+
+    def propagate_solution(self, scaled_model, original_model):
+        '''
+        This method takes the solution in scaled_model and maps it back to the original model.
+
+        It will also transform duals and reduced costs if the suffixes 'dual' and/or 'rc' are present.
+        The :code:`scaled_model` argument must be a model that was already scaled using this transformation
+        as it expects data from the transformation to perform the back mapping.
+        '''
+        if not hasattr(scaled_model, 'component_scaling_factor_map'):
+            raise AttributeError('ScaleModel:propagate_solution called with scaled_model that does not '
+                                 'have a component_scaling_factor_map. It is possible this method was called '
+                                 'using a model that was not scaled with the ScaleModel transformation')
+        component_scaling_factor_map = scaled_model.component_scaling_factor_map
+
+        # get the objective scaling factor
+        objectives = list()
+        for obj in scaled_model.component_data_objects(ctype=Objective, active=True, descend_into=True):
+            objectives.append(obj)
+        if len(objectives) != 1:
+            raise NotImplementedError('ScaleModel.propagate_solution requires a single active objective function, but %d objectives found.' % (len(objectives)))
```

Comment:
This check / exception should be done in `_apply_to` and not here.

Reply:
@jsiirola I don't see why this exception should be moved.  This is done *after* the model is transformed.  This is about the solution transformation process.

Reply:
My thought was that if the solution cannot be propagated back to the user (and the transformation will know this upfront), then the transformation should fail early, instead of waiting until after the solver spent all the time getting a solution only to tell the user "and I won't tell you what the solution is!"